File: frontend/src/types/goal.ts
export type GoalTargetType = "lift" | "body" | "frequency";
export type GoalDirection = "increase" | "decrease";

export type Goal = {
  id: string;
  userId: string;
  name: string;
  targetType: GoalTargetType;
  exerciseId: string | null;
  metric: string | null; // e.g. body metric, or "sessions_per_week"
  targetValue: string; // numeric serialized as string
  startValue: string | null; // numeric serialized as string; snapshot of where the user was when the goal was created
  unit: string;
  direction: GoalDirection;
  targetDate: string | null; // YYYY-MM-DD
  achievedAt: string | null; // ISO8601
  createdAt: string;
  updatedAt: string;
};

// Input shape for create/update. exerciseId is required for "lift", metric for "body" / "frequency".
export type GoalInput = {
  name: string;
  targetType: GoalTargetType;
  exerciseId?: string | null;
  metric?: string | null;
  targetValue: string;
  startValue?: string | null;
  unit: string;
  direction: GoalDirection;
  targetDate?: string | null;
  achievedAt?: string | null;
};

File: backend/app/models/goal.rb
class Goal < ApplicationRecord
  belongs_to :user
  belongs_to :exercise, optional: true

  TARGET_TYPES = %w[lift body frequency].freeze
  DIRECTIONS = %w[increase decrease].freeze

  validates :name, presence: true
  validates :target_type, presence: true, inclusion: { in: TARGET_TYPES }
  validates :target_value, presence: true, numericality: { greater_than: 0 }
  validates :unit, presence: true
  validates :direction, presence: true, inclusion: { in: DIRECTIONS }

  validate :type_specific_requirements

  private

  def type_specific_requirements
    case target_type
    when "lift"
      errors.add(:exercise_id, "is required for lift goals") if exercise_id.blank?
    when "body"
      errors.add(:metric, "is required for body goals") if metric.blank?
    when "frequency"
      errors.add(:metric, "is required for frequency goals") if metric.blank?
    end
  end
end
